anitya/lib/librariesio_consumer.py

```diff
@@ -0,0 +1,205 @@
+import fedmsg
+
+from anitya import get_version
+from anitya.app import SESSION as Session
+from anitya.lib import exceptions
+from anitya.lib.plugins import ECOSYSTEM_PLUGINS
+from anitya.lib.model import Project
+
+
+_log = logging.getLogger(__name__)
+
+
+class LibrariesioConsumer(FedmsgConsumer):
+    """
+    A fedmsg consumer.
+
+    Attributes:
+        topic (list): A list of strings that indicates the message topics this
+            consumer gets. This consumer accepts a single topic,
+            ``org.fedoraproject.prod.sse2fedmsg.libraryio``
+        config_key (str): A string that must be set to `True` in the fedmsg
+            configuration to enable this consumer.
+
+    """
+
+    topic = [
+        'org.fedoraproject.prod.sse2fedmsg.librariesio',
+    ]
+
+    config_key = 'anitya.libraryio.enabled'
+
+    def __init__(self, hub):
+        # If we're in development mode, rewrite some of our topics so that
+        # local playback with fedmsg-dg-replay works as expected.
+        if hub.config['environment'] == 'dev':
+            # Keep the original set, but append a duplicate set for local work
+            prefix, env = hub.config['topic_prefix'], hub.config['environment']
+            self.topic = self.topic + [
+                '.'.join([prefix, env] + topic.split('.')[3:])
+                for topic in self.topic
+            ]
+
+        super(LibrariesioConsumer, self).__init__(hub)
+
+        if not self._initialized:
+            return
+
+        # Initialize fedmsg in this thread's context.
+        hostname = socket.gethostname().split('.', 1)[0]
+        fedmsg.init(name='anitya-libraryio.%s' % hostname)
+        fedmsg.meta.make_processors(**self.hub.config)
+
+    def consume(self, message):
+        """
+        This method is called when a message with a topic this consumer
+        subscribes to arrives.
+
+        Args:
+            message (dict): The fedmsg to process.
+        """
+        project = self._fedmsg_to_project(message)
+        if project is not None:
+            # Double check the validity of the update
+            if self._validate_update(project, message):
+                session = Session()
+                session.add(project)
+                session.commit()
+                Session.remove()
+
+    def _validate_update(self, project, message):
+        """
+        Validate the project was updated.
+
+        This serves two purposes. Firstly, we connect to libraries.io over HTTP
+        so we don't have any authentication. Secondly, we might map the
+        libraries.io project to the wrong Anitya project, so this is a good
+        opportunity to catch and log those problems.
+        """
+        librariesio_version = message['body']['msg']['data']['version']
+        try:
+            anitya_version = get_version(project)
+        except exceptions.AnityaPluginException as e:
+            # TODO what to do here? Bail? Trust libraries.io?
```

Comment:
Bailing out here with a logged error is correct, as the query may fail because the candidate project was complete nonsense, and instead injected via an MitM attack against the unauthenticated feed from libraries.io.